File: #160_Intersection_of_Two_Linked_Lists/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) return null;
        ListNode tailA; int lenA = 0; ListNode tailB; int lenB = 0;
        ListNode cur = headA;
        while (cur != null) {
        	lenA++;
        	cur = cur.next;
        }
        tailA = cur;
        cur = headB;
        while(cur != null) {
        	lenB++;
        	cur = cur.next;
        }
        tailB = cur;
        if (tailB != tailA) return null;
        int diff = lenA - lenB;
        cur = (diff > 0) ? headA : headB;
        diff = Math.abs(diff);
        ListNode head = (cur == headA) ? headB : headA;
        while(diff > 0) {
        	cur = cur.next;
        	diff--;
        }
        while (cur != null && head != null) {
        	if (cur == head) return head;
        	cur = cur.next;
        	head = head.next;
        }
        return null;
    }
}

File: #160_Intersection_of_Two_Linked_Lists/Solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param two ListNodes
    # @return the intersected ListNode
    def getIntersectionNode(self, headA, headB):
        if not headA or not headB:
        	return None
        if headA is headB:
        	return headA
        la = self.lenList(headA)
        lb = self.lenList(headB)
        while la > lb:
        	headA = headA.next
        	la -= 1
        while lb > la:
        	headB = headB.next
        	lb -= 1
        while headA and headB:
        	if headA is headB:
        		return headA
        	headA = headA.next
        	headB = headB.next
        return None

    def lenList(self, head):
    	count = 0
    	while head:
    		count += 1
    		head = head.next
    	return count

sol = Solution()
h1 = ListNode(0)
n3 = ListNode(3)
n1 = ListNode(1)
h2 = ListNode(0)
h1.next = n3
n3.next = n1
h2.next = n1
print sol.getIntersectionNode(h1, h2).val